content/docs/rfcs/66/README.md

```diff
@@ -0,0 +1,39 @@
+---
+slug: 66
+title: 66/WAKU2-METADATA
+name: Waku Metadata Protocol
+status: raw
+editor: Alvaro Revuelta <alrevuelta@status.im>
+contributors:
+---
+
+# Metadata Protocol
+
+Waku specifies a req/resp protocol that provides information about the node's medatadata. Such metadata is meant to be used
+by the node to decide if a peer is worth connecting or not. The node that makes the request, includes its metadata
+so that the receiver is aware of it, without requiring an extra interaction. The parameters are the following:
+* `networkId`: Unique identifier of the network that the node is running in.
+* `shards`: Shard indexes that the node is subscribed to.
+
+
+## Protocol id
+
+`/vac/waku/metadata/1.0.0`
+
+## Request
+
+```proto
+message WakuMetadataRequest {
+  uint32 network_id = 1;
+  repeated unt32 shards = 2;
```

Comment:
@jm-clius I'm not fully aware on the implications of static sharding in here. Any thought on wether it should be considered here? Or these shards should be only for auto sharding?

In other words. If a node is running with static sharding, should the metadata protocol be able to "advertise" those shards?

Another way could be to use `string` instead of `uint32` and just put the whole topic there.

Reply:
As it stands, this would work for static sharding as well (as the cluster ID and shard numbers are encoded exactly the same). The one outstanding thing to decide is between `network_id` vs. `cluster_id` as terminology. I think there are OK arguments either side, but we should pick one and stick with it. Perhaps let's just put it up for a vote somewhere?

Interesting idea re `string`, but I see this exchange as more valuable (at least for now) for the numbered sharding as in the Waku Network. Exchanging the whole topic could as well just reuse the `hello` messages that already exist in some libp2p implementations.

Reply:
> The one outstanding thing to decide is between network_id vs. cluster_id as terminology

sure. Will ask.

> Interesting idea re string, but I see this exchange as more valuable (at least for now) for the numbered sharding as in the Waku Network

Not sure i follow. If shard is just an int, how do we know if that refers to a shard or to a static "shard"?

> Exchanging the whole topic could as well just reuse the hello messages that already exist in some libp2p implementations.

mm why then using this metadata protocol for the shards? Perhaps it makes more sense to use the hello message and the whole topics?

Reply:
> Not sure i follow. If shard is just an int, how do we know if that refers to a shard or to a static "shard"?

The network ID would be different. Static shards (despite the different name) follows the same pattern - the shard numbers are fully qualified only in conjunction with the network ID. In other words, the static sharded networks are just Waku Networks using their own network IDs (!= `1`). In fact, network IDs 0-15 are (potentially) reserved so we have some scope to define our own main/test networks, but any other network ID can be used by applications interested in creating their own Waku Networks.

> mm why then using this metadata protocol for the shards?

Well, why use it for network ID if that is also encoded in the topic? My understanding was that the intention is to conveniently exchange the minimum information that is necessary to determine if a connection is interesting or not - that would be network ID in conjunction with shard numbers. Neither of these pieces of metadata is currently enough on its own to make any final conclusions about a connection. The reason not to simply have this be implicit in a list of subscribed topics is to be explicit about our structured approach to network ID and shard numbers in topic construction. Or am I missing something behind the intention of the metadata protocol?

Reply:
> Well, why use it for network ID if that is also encoded in the topic?

mm because there are some protocols that don't know about "topics". I mean, if i dial with `store` protocol, how do i know that the store protocol is running in the same network?. afaik right now there is no way?

but yeah, agree with "exchange the minimum information to determine if a connection is interesting or not". But networkid is one level up than shards. Different network id is a deal breaker. Different shard(s), well guess we are not interested, but not as "strong".

Reply:
> I mean, if i dial with store protocol, how do i know that the store protocol is running in the same network?

True, but again I would need to know what shards within the network the store protocol supports. In all (current) cases, the network ID is just a logical grouping of shards. Since the shards (in fact the pubsub topics constructed from these) define how messages are routed, this is the minimum information I would need to know if a node is interesting for relay, store, filter and lightpush. (A future exception would be that network-ID defines the RLN contract, although this wouldn't affect my decision about which nodes I connect to or not.)